### PR TITLE
Add Request.stream

### PR DESCRIPTION
**Pull request by 38elements**
This adds `Request.stream` and `stream decorator`, as below.

```python
from sanic import Sanic
from sanic.views import CompositionView
from sanic.views import HTTPMethodView
from sanic.views import stream as stream_decorator
from sanic.blueprints import Blueprint
from sanic.response import stream, text

bp = Blueprint('blueprint_request_stream')
app = Sanic('request_stream')


class SimpleView(HTTPMethodView):

    @stream_decorator
    async def post(self, request):
        result = ''
        while True:
            body = await request.stream.get()
            if body is None:
                break
            result += body.decode('utf-8')
        return text(result)


@app.post('/stream', stream=True)
async def handler(request):
    async def streaming(response):
        while True:
            body = await request.stream.get()
            if body is None:
                break
            body = body.decode('utf-8').replace('1', 'A')
            response.write(body)
    return stream(streaming)


@bp.put('/bp_stream', stream=True)
async def bp_handler(request):
    result = ''
    while True:
        body = await request.stream.get()
        if body is None:
            break
        result += body.decode('utf-8').replace('1', 'A')
    return text(result)


async def post_handler(request):
    result = ''
    while True:
        body = await request.stream.get()
        if body is None:
            break
        result += body.decode('utf-8')
    return text(result)

app.blueprint(bp)
app.add_route(SimpleView.as_view(), '/method_view')
view = CompositionView()
view.add(['POST'], post_handler, stream=True)
app.add_route(view, '/composition_view')


if __name__ == '__main__':
    app.run(host='127.0.0.1', port=8000)
```

### CODE REVIEW

**Comment by r0fls** [comment]
does this replace https://github.com/channelcat/sanic/pull/696?

**Comment by 38elements** [comment]
I am sorry.
Yes, I do.

File: tests/test_request_stream.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+app.blueprint(bp)
+
+
+def test_request_stream():
+    data = ""
+    for i in range(1, 250000):
+        data += str(i)
+    request, response = app.test_client.post('/stream', data=data)
+    text = data.replace('1', 'A')
```
Comment (r0fls) [inline tests/test_request_stream.py]: why is this needed? 🤔

Comment (38elements) [inline tests/test_request_stream.py]: This is used [here](https://github.com/38elements/sanic/blob/4197d148e4550002fb2d351a1bb577788ba627d6/tests/test_request_stream.py#L54).

Comment (38elements) [inline tests/test_request_stream.py]: `request handler` replaces '1' with 'A'.
This makes the correct result.

Comment (38elements) [inline tests/test_request_stream.py]: I removed replacing '1' with 'A'.

File: sanic/app.py [hunk 2]
```diff
@@ -27,7 +27,7 @@
 class Sanic:
 
     def __init__(self, name=None, router=None, error_handler=None,
-                 load_env=True, request_class=None,
+                 load_env=True, request_class=None, is_request_stream=False,
```
Comment (r0fls) [inline sanic/app.py]: I think it makes more sense to set `self.is_request_stream = True` in the `app.stream` method, if used. This would avoid having to initialize it here.

Comment (r0fls) [inline sanic/app.py]: So around line 170

Comment (38elements) [inline sanic/app.py]: I think it is better to add `self.is_request_stream = True` in `app.route()` for `blueprint.stream()`. 

```
if stream:
    self.is_request_stream = True
```

**Comment by 38elements** [comment]
There is a bug.
This can not use `CompositionView`.

```
AttributeError: 'function' object has no attribute 'is_stream'
```

**Comment by 38elements** [comment]
I add `stream` in `composition_view.add()` and test for  `CompositionView` and `HTTPMethodView`.

**Comment by 38elements** [comment]
I am sorry.
I added `stream decorator` for HTTPMethodView.

**Comment by 38elements** [comment]
I think that it is necessary to put `is_request_stream` in constructor of Sanic for  `CompositionView` and `HTTPMethodView`.

**Comment by 38elements** [comment]
There is a bug.
 `CompositionView` and `HTTPMethodView`  can not set `self.is_request_stream = True`.
It is necessary to set True to `self.is_request_stream`  in `app.add_route()` for  `CompositionView` and `HTTPMethodView` .
I will fix it and add tests.

**Comment by 38elements** [comment]
I fixed it and added tests.

**Comment by seemethere** [comment]
So instead of it's own decorator wouldn't it make sense to make it as just a parameter that you can pass in through the regular method decorators?

**Comment by 38elements** [comment]
Thank you for reviewing.
Since I think it is more consistent, I think it is better `stream parameter` than `stream decorator`. 
I will delete `stream decorator` and add  `stream parameter` and change tests.

**Comment by 38elements** [comment]
I replaced stream decorator to stream parameter.
Only post, put and patch decorator have stream argument.

**Comment by seemethere** [comment]
Will merge once I create a milestone for this

**Comment by 38elements** [comment]
Thank you.

**Comment by 38elements** [comment]
I added a heading in `streaming.md`.

**Comment by r0fls** [comment]
we should also remove this from the todo here: https://github.com/channelcat/sanic/blob/master/README.rst#todo

**Comment by 38elements** [comment]
@r0fls 
I think the line should be removed by #733.

**Comment by gsvijayraajaa** [comment]
Trying to execute the code shared by @38elements . Getting the error: from sanic.views import stream as stream_decorator
ImportError: cannot import name 'stream'

Is the code merged to master ?
